### PR TITLE
Handle xjc XmlEnumValue in DynamicJAXB enum

### LINKED ISSUE #2272: DynamicJAXB doesn't handle @XmlEnumValue() from XJC class
**Describe the bug**
When XJC generated enum literals differ from the original schema, DynamicJAXB does not marshal/unmarshal correctly

XSD: xmlenum-numbers.xsd
```
<xs:schema targetNamespace="myNamespace" xmlns:myns="myNamespace" xmlns:xs="http://www.w3.org/2001/XMLSchema"
    attributeFormDefault="qualified" elementFormDefault="qualified">

    <xs:element name="tax-record">
        <xs:complexType>
            <xs:sequence>
                <xs:element name="period" type="myns:period"/>
            </xs:sequence>
        </xs:complexType>
    </xs:element>

    <!-- This test contains exactly 6 enum values to test an ASM boundary case (ICONST) -->
    <xs:simpleType name="period">
        <xs:restriction base="xs:string">
            <xs:enumeration value="QTR1"/>
            <xs:enumeration value="QTR2"/>
        </xs:restriction>
    </xs:simpleType>

</xs:schema>
```

Given the above XSD, XJC will generated an enum like below
```
public class enum Period {
@XmlEnumValue("QTR1")
QTR_1("QTR1")
@XmlEnumValue("QTR2")
QTR_2("QTR2");

private final String value;

Period(String v) {
  value = v
};
}
```

When unmarshalling the below Xml, the `Period` enum is not being evaluated by Dynamic JAXBContext
```
<tax-record>
<period>QTR1</period>
</tax-record>
```

`period` attribute will show as `null`

**To Reproduce**
Steps/resources to reproduce the behavior:

- EclipseLink version 4.0.4
- Java/JDK version 17.0.11

**Expected behavior**
When marshalling enum with a differing 'value' from the literal, the value should be used rather than the literal.

XML
```
<tax-record>
<period>QTR1</period>
</tax-record>
```

Unmarshals to:
TaxRecord:
- period: Period.QTR_1

And will marshal back into the same XML as above

**Additional context**
Below is a test case which should pass

```
public void testXmlEnumWithNumbers() throws Exception {
        // Tests XmlEnum and XmlEnumValue

        InputStream inputStream = ClassLoader.getSystemResourceAsStream(RESOURCE_DIR + "xmlenum-numbers.xsd");
        jaxbContext = DynamicJAXBContextFactory.createContextFromXSD(inputStream, null, null, null);

        DynamicEntity taxRecord = jaxbContext.newDynamicEntity(PACKAGE + ".TaxRecord");
        assertNotNull("Could not create Dynamic Entity.", taxRecord);

        Object QTR_1 = jaxbContext.getEnumConstant(PACKAGE + ".Period", "QTR_1");
        assertNotNull("Could not find enum constant.", QTR_1);

        taxRecord.set("period", QTR_1);

        Document marshalDoc = DocumentBuilderFactory.newInstance().newDocumentBuilder().newDocument();
        jaxbContext.createMarshaller().marshal(taxRecord, marshalDoc);

        assertEquals("QTR1", marshalDoc.getDocumentElement().getTextContent());
    }

```

### PR DESCRIPTION
**Pull request by timtatt**
To resolve #2272

- Changes EnumInfo literals from List to Map to store intended value
- Adds @XmlEnumValue annotation to DynamicJAXB enum class
- Check for @XmlEnumValue when adding enum conversion values

### CODE REVIEW

File: foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/dynamic/DynamicClassLoader.java [hunk 1]
```diff
@@ -29,10 +29,8 @@
 import org.eclipse.persistence.internal.security.PrivilegedAccessHelper;
 import org.eclipse.persistence.sessions.Session;
 
-import java.util.ArrayList;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
```
Comment (rfelcman) [inline foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/dynamic/DynamicClassLoader.java:32]: Please don't use `*` in imports. We prefer specific imports like removed part.

File: foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/dynamic/DynamicClassLoader.java [hunk 2]
```diff
@@ -96,17 +94,20 @@ public EclipseLinkClassWriter getClassWriter(String className) {
     }
 
     public void addEnum(String className, Object... literalLabels) {
+        var literalsMap = Arrays.stream(literalLabels)
```
Comment (rfelcman) [inline foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/dynamic/DynamicClassLoader.java:97]: `var` is not preferred as we will backport this PR into 2.7 branch with JDK 8.0 support.

File: moxy/org.eclipse.persistence.moxy/src/test/resources/org/eclipse/persistence/testing/jaxb/dynamic/xmlenum-numbers.xsd [hunk 3]
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+
+    Copyright (c) 2018, 2024 Oracle and/or its affiliates. All rights reserved.
```
Comment (rfelcman) [inline moxy/org.eclipse.persistence.moxy/src/test/resources/org/eclipse/persistence/testing/jaxb/dynamic/xmlenum-numbers.xsd:4]: For a new files `Copyright (c) 2024 Oracle...` is correct.

File: moxy/org.eclipse.persistence.moxy/src/test/resources/org/eclipse/persistence/testing/jaxb/dynamic/xmlenum-numbers.xsd [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+    <xs:element name="tax-record">
+        <xs:complexType>
+            <xs:sequence>
+                <xs:element name="period" type="myns:period"/>
+            </xs:sequence>
+        </xs:complexType>
+    </xs:element>
+
+    <!-- This test contains exactly 6 enum values to test an ASM boundary case (ICONST) -->
```
Comment (rfelcman) [inline moxy/org.eclipse.persistence.moxy/src/test/resources/org/eclipse/persistence/testing/jaxb/dynamic/xmlenum-numbers.xsd:27]: This comment seems inconsistent with values `6 enum` vs 4 (`QTR1...QTR4`).

File: moxy/org.eclipse.persistence.moxy/src/main/java/org/eclipse/persistence/jaxb/dynamic/metadata/SchemaMetadata.java [hunk 5]
```diff
@@ -17,17 +17,17 @@
 import java.io.ByteArrayInputStream;
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
-import java.util.ArrayList;
-import java.util.HashSet;
-import java.util.Iterator;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
```
Comment (rfelcman) [inline moxy/org.eclipse.persistence.moxy/src/main/java/org/eclipse/persistence/jaxb/dynamic/metadata/SchemaMetadata.java:20]: Please use specific imports instead of `... .*` This should be applied to all changes.

File: foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/dynamic/DynamicClassWriter.java [hunk 6]
```diff
@@ -223,7 +223,7 @@ protected void addMethods(ClassWriter cw, String parentClassType) {
 
     protected byte[] createEnum(EnumInfo enumInfo) {
 
-        String[] enumValues = enumInfo.getLiteralLabels();
+        var enumValues = enumInfo.getEnumValues();
```
Comment (rfelcman) [inline foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/dynamic/DynamicClassWriter.java:226]: Please use specific type instead of `var`.

File: foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/dynamic/DynamicClassWriter.java [hunk 7]
```diff
@@ -232,8 +232,10 @@ protected byte[] createEnum(EnumInfo enumInfo) {
         cw.visit(Opcodes.ACC_PUBLIC + Opcodes.ACC_FINAL + Opcodes.ACC_SUPER + Opcodes.ACC_ENUM, internalClassName, null, "java/lang/Enum", null);
 
         // Add the individual enum values
-        for (String enumValue : enumValues) {
-            cw.visitField(Opcodes.ACC_PUBLIC + Opcodes.ACC_FINAL + Opcodes.ACC_STATIC + Opcodes.ACC_ENUM, enumValue, "L" + internalClassName + ";", null, null);
+        for (var enumValue : enumValues.entrySet()) {
+            var fv = cw.visitField(Opcodes.ACC_PUBLIC + Opcodes.ACC_FINAL + Opcodes.ACC_STATIC + Opcodes.ACC_ENUM, enumValue.getKey(), "L" + internalClassName + ";", null, null);
```
Comment (rfelcman) [inline foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/dynamic/DynamicClassWriter.java:236]: `var` again.

**Review by rfelcman — COMMENTED** [review]
In general please use specific types instead of `var`.

File: moxy/org.eclipse.persistence.moxy/src/main/java/org/eclipse/persistence/jaxb/JAXBEnumTypeConverter.java [hunk 8]
```diff
@@ -92,6 +95,16 @@ public void initialize(DatabaseMapping mapping, Session session) {
         super.initialize(mapping, session);
     }
 
+    private String getEnumValue(Enum theEnum) {
+        try {
+            Field field = theEnum.getClass().getField(theEnum.name());
```
Comment (rfelcman) [inline moxy/org.eclipse.persistence.moxy/src/main/java/org/eclipse/persistence/jaxb/JAXBEnumTypeConverter.java:100]: Please use (maybe add new method into) https://github.com/eclipse-ee4j/eclipselink/blob/master/moxy/org.eclipse.persistence.moxy/src/main/java/org/eclipse/persistence/jaxb/ReflectionUtils.java
to ensure, that `PrivilegedAccessHelper.callDoPrivilegedWithException(` is used instead of direct call.

**Review by rfelcman — APPROVED** [review]
LGTM

**Comment by rfelcman** [comment]
It was merged into master -> upcoming 5.x version. If You want it into EL 4.x please make a backport against 4.0 branch.

**Comment by timtatt** [comment]
@rfelcman thanks for reviewing. Is it possible to kick off a beta promotion to publish to mvnrepository?

**Comment by rfelcman** [comment]
There is snapshot repo
https://jakarta.oss.sonatype.org/content/repositories/staging/org/eclipse/persistence/eclipselink/5.0.0-SNAPSHOT/
Every new snapshot is build after PR merge (nightly).
5.0.0-B04 is planned for a tomorrow,
